components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -396,12 +399,47 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
             _agentConfig.TriggerConfigUpdate(this, new EventArgs());
         }
 
+        private static void AdjustResponse(ref List<string> statusResponse) {
+            byte[] result;
+            int maxIndex = statusResponse.Count();
+            string resultString = string.Empty;
+            while (true) {
+                resultString = JsonConvert.SerializeObject(statusResponse.GetRange(0, maxIndex));
+                result = Encoding.UTF8.GetBytes(resultString);
+                if (result.Length > MaxResponsePayloadLength) {
+                    maxIndex /= 2;
+                    continue;
+                }
+                else {
+                    break;
+                }
+            }
+            if (maxIndex != statusResponse.Count()) {
+                statusResponse.RemoveRange(maxIndex, statusResponse.Count() - maxIndex);
+                statusResponse.Add("Results have been cropped due to package size limitations.");
+            }
+        }
+        private MethodResponse BuildResponse(List<string> statusResponse, HttpStatusCode statusCode, string prefix) {
+            AdjustResponse(ref statusResponse);
+            string resultString = JsonConvert.SerializeObject(statusResponse);
+            byte[] result = Encoding.UTF8.GetBytes(resultString);
+            if (result.Length > MaxResponsePayloadLength) {
+                _logger.Error($"{prefix} Response size is too long");
+                Array.Resize(ref result, result.Length > MaxResponsePayloadLength ? MaxResponsePayloadLength : result.Length);
+            }
+            MethodResponse methodResponse = new MethodResponse(result, (int)statusCode);
```

Comment:
The result encoding is handled by the MethodRouter class. We just need to pass the response as an json encodable object. I believe the method router even has logic to handle the continuation tokens, we should investigate if that is enough for our needs.